File: iOS/Kalorie/Core/UseCases/DeleteMySubmissionUseCase.swift
//
//  DeleteMySubmissionUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 15.09.2026.
//

import Foundation

protocol DeleteMySubmissionUseCaseProtocol {
    func callAsFunction(id: String) async throws
}

struct DeleteMySubmissionUseCase: DeleteMySubmissionUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(id: String) async throws {
        guard authProvider.userId != nil else { throw AuthError.notAuthenticated }
        try await dataProvider.deleteAsync(id: id, from: Constants.Firestore.foodItemSubmissions)
    }
}

#if DEBUG
struct DeleteMySubmissionUseCaseFake: DeleteMySubmissionUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(id: String) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/DeleteMySubmissionUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.utils.Constants

interface DeleteMySubmissionUseCaseProtocol {
    suspend operator fun invoke(id: String)
}

class DeleteMySubmissionUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : DeleteMySubmissionUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(id: String) {
        if (authProvider.userId == null) throw AuthError.NotAuthenticated
        dataProvider.deleteAsync(id = id, from = Constants.Firestore.FOOD_ITEM_SUBMISSIONS)
    }
}
